src/tree/v2/providers/CompatibleBranchDataProvider.ts

```diff
@@ -0,0 +1,72 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { AzExtParentTreeItem, AzExtTreeDataProvider, AzExtTreeItem, IFindTreeItemContext, ITreeItemPickerContext } from '@microsoft/vscode-azext-utils';
+import { AppResource, AppResourceResolver } from '@microsoft/vscode-azext-utils/hostapi';
+import * as vscode from 'vscode';
+import { ApplicationResource, BranchDataProvider, ResourceModelBase } from '../../../api/v2/v2AzureResourcesApi';
+
+export class CompatibleBranchDataProvider<TResource extends ApplicationResource, TModel extends AzExtTreeItem & ResourceModelBase> extends AzExtTreeDataProvider implements BranchDataProvider<TResource, TModel> {
+    private readonly overrideOnDidChangeTreeDataEmitter = new vscode.EventEmitter<TModel | undefined>();
+
+    public constructor(private readonly resolver: AppResourceResolver, loadMoreCommandId: string) {
+        super(undefined as unknown as AzExtParentTreeItem, loadMoreCommandId);
+    }
+
+    //#region TreeDataProvider
+
+    public override readonly onDidChangeTreeData = this.overrideOnDidChangeTreeDataEmitter.event;
+
+    public override getParent(treeItem: TModel): Promise<TModel> {
+        // This method is redeclared to make TypeScript happier, but it's no more than a super call with extra casts
+        return super.getParent(treeItem) as Promise<TModel>;
+    }
+
+    public override getChildren(treeItem: TModel & AzExtParentTreeItem): Promise<TModel[]> {
+        // This method is redeclared to make TypeScript happier, but it's no more than a super call with extra casts
+        return super.getChildren(treeItem) as Promise<TModel[]>;
+    }
+
+    //#endregion
+
+    //#region BranchDataProvider
+
+    public async getResourceItem(element: TResource): Promise<TModel> {
+        const oldAppResource: AppResource = {
+            ...element,
+            type: element.type.type,
+            kind: element.type.kinds?.join(';'),
+        };
+
+        return this.resolver.resolveResource(/* TODO */'foo', oldAppResource) as Promise<TModel>;

```

Comment:
Do we have a plan for how subscriptions will be passed down yet?

Reply:
@philliphoff has been playing around with the VSCode auth API, he could tell you more.

Reply:
Right now I'm working toward having application resources contain a subscription property (like they do today), where the subscription object contains a `getSession()` method that returns a VS Code authentication session.  Whether the session is an *actual* VS Code authentication session or just a wrapper around the existing Azure Account API is to be seen. It'll likely start as the latter, but this allows us to eventually move to VS Code's authentication system under the covers (e.g. when ADAL/MSAL limitations get resolved, etc.).  This also insulates our API from T1, T2, ..., Tn credentials changes.  That is, the intent is that the Azure Utils library would have convenience methods to convert a VS Code `AuthenticationSession` to a `TokenCredential` or whatever the Azure SDKs need in the future, without requring us to update our API.

```typescript
export interface ApplicationAuthentication {
    getSession(scopes?: string[]): vscode.ProviderResult<vscode.AuthenticationSession>;
}

/**
 * Information specific to the Subscription
 */
export interface ApplicationSubscription {
    readonly authentication: ApplicationAuthentication;
    readonly displayName: string;
    readonly subscriptionId: string;
    readonly environment: Environment;
    readonly isCustomCloud: boolean;
}

/**
 * Represents an individual resource in Azure.
 * @remarks The `id` property is expected to be the Azure resource ID.
 */
export interface ApplicationResource extends ResourceBase {
    readonly subscription: ApplicationSubscription;
    readonly type: ApplicationResourceType;
    readonly location?: string;
    readonly resourceGroup?: string;
    /** Resource tags */
    readonly tags?: {
        [propertyName: string]: string;
    };
    /* add more properties from GenericResource if needed */
}
```

Note that we don't just expose the `AuthenticationSession`, even though that should generally work, instead having an intermediate `getSession()` function.  This is because of the way `TokenCredential.getToken()` works where new scopes may be injected, whereas VS Code tends to expect that scopes are all defined up front.  That said, we *generally* know what scopes will be needed, so we can "pre-authenticate" with VS Code with those scopes and keep prompts to a minimum.